metals/src/main/scala/scala/meta/internal/metals/FormattingProvider.scala

```diff
@@ -0,0 +1,93 @@
+package scala.meta.internal.metals
+
+import scala.meta._
+import org.eclipse.{lsp4j => l}
+import MetalsEnrichments._
+import scala.language.reflectiveCalls
+import java.util
+import java.nio.file.Files
+import scala.util.Try
+
+/**
+ * Implement text formatting using Scalafmt
+ */
+final class FormattingProvider(
+    workspace: AbsolutePath,
+    buffers: Buffers,
+    embedded: Embedded,
+    userConfig: () => UserConfiguration
+) {
+
+  def format(path: AbsolutePath): util.List[l.TextEdit] = {
+    val input = path.toInputFromBuffers(buffers)
+    val fullDocumentRange = Position.Range(input, 0, input.chars.length).toLSP
+    (for {
+      scalafmt <- classloadedScalafmt
+      formatted <- scalafmtConfigPath match {
+        case Some(configPath) =>
+          scalafmt
+            .format(input.text, configPath.toString, input.path)
+            .fold(e => {
+              scribe.error("Error while running Scalafmt", e)
+              None
+            }, Some(_))
+        case None if !userConfig().scalafmtRequireConfigFile =>
+          scalafmt
+            .format(input.text, input.path)
+            .fold(e => {
+              scribe.error("Error while running Scalafmt", e)
```

Comment:
These errors may be noisy, since they are triggered even when trying to format a non-parsing buffer. I'm open to suggestions.